bindings/ergo-lib-c/src/macros.rs

```diff
@@ -0,0 +1,133 @@
+#[allow(unused_imports)]
+use paste::paste;
+
+/// `ergo-lib` uses a number of collection types that are simple wrappers around `Vec`. We have a
+/// generic type for such a collection in `ergo-lib-c-core::collections::Collection`. A limitation
+/// of `cbindgen` is that it cannot process generic functions. This macro generates a C-compatible
+/// interface for such a collection for any desired type T.
+///
+/// As an example the call `make_collection(BlockHeaders, BlockHeader);` generates:
+///
+///```
+///pub type BlockHeadersPtr = CollectionPtr<BlockHeader>;
+///pub type ConstBlockHeadersPtr = ConstCollectionPtr<BlockHeader>;
+///
+///#[no_mangle]
+///pub unsafe extern "C" fn ergo_wallet_block_headers_new(collection_ptr_out: *mut BlockHeadersPtr) -> ErrorPtr {
+///    let res = collection_new(collection_ptr_out);
+///    Error::c_api_from(res)
+///}
+///
+///#[no_mangle]
+///pub unsafe extern "C" fn ergo_wallet_block_headers_delete(collection_ptr_out: BlockHeadersPtr) {
+///    collection_delete(collection_ptr_out)
+///}
+///
+///#[no_mangle]
+///pub unsafe extern "C" fn ergo_wallet_block_headers_len(
+///    collection_ptr: ConstBlockHeadersPtr,
+///) -> ReturnNum<usize> {
+///    match collection_len(collection_ptr) {
+///        Ok(value) => ReturnNum {
+///            value: value as usize,
+///            error: std::ptr::null_mut(),
+///        },
+///        Err(e) => ReturnNum {
+///            value: 0, // Just a dummy value
+///            error: Error::c_api_from(Err(e)),
+///        },
+///    }
+///}
+///
+///#[no_mangle]
+///pub unsafe extern "C" fn ergo_wallet_block_headers_get(
+///    collection_ptr: ConstBlockHeadersPtr,
+///    index: usize,
+///    element_ptr_out: *mut BlockHeaderPtr,
+///) -> ErrorPtr {
+///    match collection_get(collection_ptr, index) {
+///        Ok(Some(bh)) => {
+///            *element_ptr_out = Box::into_raw(Box::new(bh));
+///            Error::c_api_from(Ok(()))
+///        }
+///        Ok(None) => Error::c_api_from(Err(Error::Misc("BlockHeaders: index out of bounds".into()))),
+///        Err(e) => Error::c_api_from(Err(e)),
+///    }
+///}
+///
+///#[no_mangle]
+///pub unsafe extern "C" fn ergo_wallet_block_headers_add(
+///    element_ptr: ConstBlockHeaderPtr,
+///    collection_ptr_out: BlockHeadersPtr,
+///) -> ErrorPtr {
+///    Error::c_api_from(collection_add(collection_ptr_out, element_ptr))
+///}
+///```
+#[macro_export]
+macro_rules! make_collection {
```

Comment:
Neat!

Reply:
Thanks. I think it's worth the trouble of using nightly for `cbindgen` to generate this boilerplate.

Reply:
Oh yeah,  I agree.